server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -0,0 +1,366 @@
+     * @throws IOException if the file exists but could not be deleted.
+     */
+    @Override
+    public void deleteFile(String name) throws IOException {
+        String remoteFilename = getExistingRemoteFilename(name);
+        if (remoteFilename != null) {
+            remoteDataDirectory.deleteFile(remoteFilename);
+            segmentsUploadedToRemoteStore.remove(name);
+        }
+    }
+
+    /**
+     * Returns the byte length of a segment file in the remote segment store.
+     * @param name the name of an existing segment file in local filesystem.
+     * @throws IOException in case of I/O error
+     * @throws NoSuchFileException if the file does not exist in the cache or remote segment store
+     */
+    @Override
+    public long fileLength(String name) throws IOException {
+        String remoteFilename = getExistingRemoteFilename(name);
+        if (remoteFilename != null) {
+            return remoteDataDirectory.fileLength(remoteFilename);
+        } else {
+            throw new NoSuchFileException(name);
+        }
+    }
+
+    /**
+     * Creates and returns a new instance of {@link RemoteIndexOutput} which will be used to copy files to the remote
+     * segment store.
+     * @param name the name of the file to create.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
+        return remoteDataDirectory.createOutput(getNewRemoteSegmentFilename(name), context);
+    }
+
+    /**
+     * Opens a stream for reading an existing file and returns {@link RemoteIndexInput} enclosing the stream.
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     * @throws NoSuchFileException if the file does not exist either in cache or remote segment store
+     */
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        String remoteFilename = getExistingRemoteFilename(name);
+        if (remoteFilename != null) {
+            return remoteDataDirectory.openInput(remoteFilename, context);
+        } else {
+            throw new NoSuchFileException(name);
+        }
+    }
+
+    /**
+     * Copies an existing src file from directory from to a non-existent file dest in this directory.
+     * Once the segment is uploaded to remote segment store, update the cache accordingly.
+     */
+    @Override
+    public void copyFrom(Directory from, String src, String dest, IOContext context) throws IOException {
+        String remoteFilename = getNewRemoteSegmentFilename(dest);
+        remoteDataDirectory.copyFrom(from, src, remoteFilename, context);
+        String checksum = getChecksumOfLocalFile(from, src);
+        UploadedSegmentMetadata segmentMetadata = new UploadedSegmentMetadata(src, remoteFilename, checksum);
+        segmentsUploadedToRemoteStore.put(src, segmentMetadata);
+    }
```

Comment:
Maybe we need to pass the computed checksum to remote store when we upload too?

Reply:
We do that as part of metadata upload. Each line in metadata file would be: original filename, uploaded filename, checksum